src/sqlfluff/core/dialects/base.py

```diff
@@ -152,7 +152,58 @@ def replace(self, **kwargs: DialectElementType):
         for n in kwargs:
             if n not in self._library:  # pragma: no cover
                 raise ValueError(f"{n!r} is not already registered in {self!r}")
-            self._library[n] = kwargs[n]
+            cls = kwargs[n]
+            if self._library[n] is cls:
+                continue
+
+            # To replace a segment, the replacement must either be a
+            # subclass of the original, *or* it must have the same
+            # public methods and/or fields as it.
+            base_dir = set(dir(self._library[n]))
+            subclass = False
+            if isinstance(self._library[n], type) and isinstance(cls, type):
+                subclass = issubclass(cls, self._library[n])
+                if not subclass:
+                    if self._library[n].type != cls.type:
+                        raise ValueError(  # pragma: no cover
+                            f"Cannot replace {n!r} because 'type' property does not "
+                            f"match: {cls.type} != {self._library[n].type}"
+                        )
+
+                    cls_dir = set(dir(cls))
+                    missing = set(
+                        n for n in base_dir.difference(cls_dir) if not n.startswith("_")
+                    )
+                    if missing:
+                        raise ValueError(  # pragma: no cover
+                            f"Cannot replace {n!r} because it's not a subclass and "
+                            f"is missing these from base: {', '.join(missing)}"
+                        )
+
+            if subclass:
+                # If the segment class we're replacing defines these fields, the
+                # replacement must override either:
+                # - NONE of them or
+                # - ALL of them
+                # Overriding a subset of them is not necessarily wrong, but it's
+                # error-prone, hence this policy.
+                grammars = {"match_grammar", "parse_grammar"}
+                # TRICKY: The explicit use of __dict__ on the classes is
+                # deliberate. We are concerned with whether a class itself does
+                # or does not define a thing, IGNORING INHERITED VALUES.
+                if grammars.intersection(set(self._library[n].__dict__)) == grammars:
+                    overrides = grammars.intersection(set(cls.__dict__))
+                    if overrides and overrides != grammars:
+                        for grammar in grammars:  # pragma: no cover
```

Comment:
I presume this `no cover` can be removed now?

```suggestion
                        for grammar in grammars:
```